File: Disjoint Set & Heap/jichon/20040.js
const fs = require("fs");
const input = fs.readFileSync("./input.txt").toString().trim().split("\n");

// disjoint set -> 서로소 집합
// union-find
// find : 집합 찾기 -> 루트 노드까지 끝까지 찾음
const find_parent = (parent, x) => {
  if (parent[x] !== x) {
    parent[x] = find_parent(parent, parent[x]);
  }
  return parent[x];
};

// union : 두 원소가 속한 집합 합치기
const union_parent = (parent, a, b) => {
  a = find_parent(parent, a);
  b = find_parent(parent, b);
  if (a < b) parent[b] = a;
  else parent[a] = b;
};

const solution = () => {
  const [N, M] = input[0].split(" ").map(Number);
  const parent = Array.from({ length: N + 1 }, (_, i) => i);

  let isCycle = false;
  let cyclenum = 0;

  for (let i = 1; i <= M; i++) {
    const [a, b] = input[i].split(" ").map(Number);
    // root가 똑같다 -> 돌고돈다는 말 아닌가!
    if (find_parent(parent, a) === find_parent(parent, b)) {
      isCycle = true;
      cyclenum = i;
      break;
    } else {
      union_parent(parent, a, b);
    }
  }

  console.log(cyclenum);
};

solution();

File: Disjoint Set/20040.py
import sys
from collections import deque

input = sys.stdin.readline
n,m = map(int, input().split())
parent = []
result = 0

def find(x):
    if x != parent[x]:
        parent[x] = find(parent[x])
    
    return parent[x]
    
def union(a,b):
    # 집합이 같으면 사이클 아니면 합치기
    a = find(a)
    b = find(b)
    
    if a == b:
        return False 
    parent[b] = a
    return True
        

# parent = list(range(n))
for i in range(n):
    parent.append(i)

for i in range(1,m+1):
    a, b = map(int, input().split())
    if union(a,b) == 0:
        result = i
        break
else:
    print(0)

print(result)